File: MoonEngine-Root/OpenGL/src/main/java/com/greentree/engine/moon/opengl/adapter/buffer/PushCommandBuffer.java
package com.greentree.engine.moon.opengl.adapter.buffer;

import com.greentree.commons.graphics.smart.mesh.Mesh;
import com.greentree.commons.graphics.smart.shader.material.Material;
import com.greentree.commons.image.Color;
import com.greentree.engine.moon.opengl.adapter.GLRenderLibrary;
import com.greentree.engine.moon.opengl.adapter.OpenGLMaterial;
import com.greentree.engine.moon.opengl.adapter.RenderMesh;
import com.greentree.engine.moon.opengl.adapter.buffer.command.*;
import com.greentree.engine.moon.render.pipeline.target.buffer.TargetCommandBuffer;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class PushCommandBuffer implements TargetCommandBuffer {

    private static final int INITIAL_CAPACITY = 10;
    private final GLRenderLibrary library;
    private final List<TargetCommand> commands;
    private boolean depthTest;
    private boolean cullFace;
    private OpenGLMaterial material;
    private RenderMesh mesh;

    public PushCommandBuffer(GLRenderLibrary library) {
        this(library, INITIAL_CAPACITY);
    }

    public PushCommandBuffer(GLRenderLibrary library, int initialCapacity) {
        this.library = library;
        commands = new ArrayList<>(initialCapacity);
    }

    @Override
    public void clear() {
        commands.clear();
        cullFace = false;
        depthTest = false;
    }

    @Override
    public void clearColor(Color color) {
        push(new ClearRenderTargetColor(color));
    }

    @Override
    public void clearDepth(float depth) {
        push(new ClearRenderTargetDepth(depth));
    }

    @Override
    public void disableCullFace() {
        cullFace = false;
    }

    @Override
    public void disableDepthTest() {
        depthTest = false;
    }

    @Override
    public void bindMesh(Mesh mesh) {
        this.mesh = (RenderMesh) mesh;
    }

    @Override
    public void bindMaterial(Material material) {
        this.material = (OpenGLMaterial) material;
    }

    @Override
    public void draw() {
        TargetCommand command = new DrawMesh(mesh, material);
        if (cullFace)
            command = new CullFace(command);
        if (depthTest)
            command = new DepthTest(command);
        push(command);
    }

    @Override
    public void enableCullFace() {
        cullFace = true;
    }

    @Override
    public void enableDepthTest() {
        depthTest = true;
    }

    @Override
    public void execute() {
        for (var c : commands)
            c.run();
    }

    public void push(TargetCommand command) {
        Objects.requireNonNull(command);
        while (commands.size() > 1) {
            final var c = commands.get(commands.size() - 1);
            final var m = c.merge(command);
            if (m == null)
                break;
            command = m;
            commands.remove(commands.size() - 1);
        }
        commands.add(command);
    }

}

File: MoonEngine-Root/OpenGL/src/main/java/com/greentree/engine/moon/opengl/adapter/buffer/PushCommandBuffer.kt
package com.greentree.engine.moon.opengl.adapter.buffer

import com.greentree.commons.graphics.smart.mesh.Mesh
import com.greentree.commons.graphics.smart.shader.material.Material
import com.greentree.commons.graphics.smart.target.RenderCommandBuffer
import com.greentree.commons.image.Color
import com.greentree.engine.moon.opengl.adapter.GLRenderContext
import com.greentree.engine.moon.opengl.adapter.OpenGLMaterial
import com.greentree.engine.moon.opengl.adapter.RenderMesh
import com.greentree.engine.moon.opengl.adapter.buffer.command.ClearRenderTargetColor
import com.greentree.engine.moon.opengl.adapter.buffer.command.ClearRenderTargetDepth
import com.greentree.engine.moon.opengl.adapter.buffer.command.CullFace
import com.greentree.engine.moon.opengl.adapter.buffer.command.DepthTest
import com.greentree.engine.moon.opengl.adapter.buffer.command.DrawMesh
import com.greentree.engine.moon.opengl.adapter.buffer.command.TargetCommand

class PushCommandBuffer @JvmOverloads constructor(
	private val library: GLRenderContext,
	initialCapacity: Int = INITIAL_CAPACITY,
) : RenderCommandBuffer {

	private val commands: MutableList<TargetCommand>
	private var depthTest = false
	private var cullFace = false
	private var material: OpenGLMaterial? = null
	private var mesh: RenderMesh? = null
	private var optimazed = true

	init {
		commands = ArrayList(initialCapacity)
	}

	override fun bindMaterial(material: Material) {
		this.material = material as OpenGLMaterial
	}

	override fun bindMesh(mesh: Mesh) {
		this.mesh = mesh as RenderMesh
	}

	override fun clear() {
		commands.clear()
		cullFace = false
		depthTest = false
	}

	override fun clearColor(color: Color) {
		push(ClearRenderTargetColor(color))
	}

	override fun clearDepth(depth: Float) {
		push(ClearRenderTargetDepth(depth))
	}

	override fun disableCullFace() {
		cullFace = false
	}

	override fun disableDepthTest() {
		depthTest = false
	}

	override fun draw(count: Int) {
		var command: TargetCommand = DrawMesh(mesh!!, material!!, count)
		if(cullFace) command = CullFace(command)
		if(depthTest) command = DepthTest(command)
		push(command)
	}

	override fun enableCullFace() {
		cullFace = true
	}

	override fun enableDepthTest() {
		depthTest = true
	}

	override fun execute() {
		if(!optimazed) {
			optimazed = true
//			val cs = ArrayList<TargetCommand>(commands.size)
//			var c: Int
//			do {
//				c = 0
//				var i = 0
//				while(i < commands.size - 1) {
//					val a = commands[i]
//					val b = commands[i + 1]
//					val m = a.merge(b)
//					if(m == null) {
//						cs.add(a)
//					} else {
//						cs.add(m)
//						i++
//						c++
//					}
//					i++
//				}
//				commands.clear()
//				commands.addAll(cs)
//				cs.clear()
//			} while(c > 0)
//
		}
		for(c in commands) c.run()
	}

	private fun push(command: TargetCommand) {
//		optimazed = false
		var command = command
		while(commands.size > 1) {
			val c = commands[commands.size - 1]
			val m = c.merge(command) ?: break
			command = m
			commands.removeAt(commands.size - 1)
		}
		commands.add(command)
	}

	companion object {

		private const val INITIAL_CAPACITY = 10
	}
}